add a 'unified palette' option to a scene

cycomachead opened the pull request:
/cc @ddgarcia @maryfries @pamelafox

The primary driving factor behind this is the middle school course, with the intention most blocks will be hidden. It would be great if we could get this in for snapcon. 😀 

Adds a new menu option, local to each scene and saved in the XML, called "Unified Palette". 
Clicking on a category selector scrolls the palette to that location. 

In a unified palette, there is only 1 make a block button. 
It seemed moderately weird to have one button after every category -- especially since in our use case most blocks will be hidden.
Each category's primitives are followed by the respective custom blocks. 


<img width="230" alt="Screen Shot 2021-06-29 at 8 09 13 PM" src="https://user-images.githubusercontent.com/1505907/123896210-f3a2b780-d915-11eb-895c-23a3e21fa27a.png">
<img width="1792" alt="Screen Shot 2021-06-29 at 8 09 02 PM" src="https://user-images.githubusercontent.com/1505907/123896211-f43b4e00-d915-11eb-8fdf-570b38c44dc8.png">


**Code Notes:**
Along the path of implementing this, I originally modified `blockTemplates`, which initially seemed like the most logical place, but I think the current solution is better. However, I kept some refactoring around which while not necessary, I think makes `blockTemplates` more readable. I also renamed `blocksCache` to `primitivesCache`.

brianharvey commented:
Two questions:

1.  Are all the options in the bottom section of the Settings menu becoming scene-specific?  If not, if some remain entire-project-specific, maybe we should add another hrule before the scene-specific ones.

2.  Can we take advantage of this code to implement the feature of showing all (visible) custom blocks in the Other palette view (as well as where they belong by color), even when Unified Palette isn't on?  Seems like you've done most of the work already.

cycomachead commented:
Test URL: https://snap.berkeley.edu/versions/michael-dev/snap.html

> * Are all the options in the bottom section of the Settings menu becoming scene-specific?  If not, if some remain entire-project-specific, maybe we should add another hrule before the scene-specific ones.

I think so, but haven't tested them all.

> 2\. Can we take advantage of this code to implement the feature of showing all (visible) custom blocks in the Other palette view (as well as where they belong by color), even when Unified Palette isn't on?  Seems like you've done most of the work already.

We already said we'd wait. But, only sort of... Mainly, we _don't have_ an Other palette still... just a Variables one.  There's too much unsettled design here.

brianharvey commented:
I don't really see why the current merging of Vars/Lists/Other into one column, bad as that is, stands in the way of adding (all visible) custom blocks to Other.  As for "unsettled design," do any of us think we *shouldn't* separate those three?  At one point Jens was opposed because he strongly wanted the category list not to shrink the space for the palette itself, but I think in light of wanting to support extensions with additional categories he's come around.

RubenVerg commented:
@brianharvey while I definitely support splitting Other from Variables/Lists, the "missing category" could also take the space of a program flow/events & messages split, or maybe Remote Controlling, empty but meant for custom blocks like for controlling bots or whatever, might also be interesting to support web sockets there, or do split vars and lists but calling the List category "Data" or "Data Structures", since it's common to implement other structures (most notably dictionaries/maps) in the Lists category.

jmoenig commented:
This is really cool, I love it!

cycomachead commented:
Thanks, Jens!


Can we _please_ discuss the "all custom blocks" view and acceptable list of categories in its own issue?

I'd suggest #908 (which probably should be redone rather than resolving conflicts since the changes are small).

> I don't really see why the current merging of Vars/Lists/Other into one column, bad as that is, stands in the way of adding (all visible) custom blocks to Other.

What I am saying is that there is no individual "Other" palette. This PR doesn't change that... We could shove all custom blocks in the "other section" of the current variables palette...but that seems wrong and messy, and not really what people want.

jmoenig commented:
I'm fine with pulling this before SnapCon, Michael.

I'm really annoyed by the discussion about an "all custom blocks view", and especially about listing custom blocks twice in a separate category, so I won't participate in it.

cycomachead commented on src/gui.js:
```diff
@@ -3609,7 +3640,7 @@ IDE_Morph.prototype.settingsMenu = function () {
             */
             Process.prototype.enableJS = !Process.prototype.enableJS;
-            this.currentSprite.blocksCache.operators = null;
+            this.currentSprite.primitivesCache.operators = null;
```
Technically speaking these were not necessary changes, but seemed more clear. (Since I was originally trying to stuff custom blocks cache in here, and that actually got a little messy...)

cycomachead commented on src/gui.js:
```diff
@@ -5533,6 +5571,14 @@ IDE_Morph.prototype.switchToScene = function (scene, refreshAlbum) {
     scene.applyGlobalSettings();
     this.world().keyboardFocus = this.stage;
+
+    if (this.currentCategory != 'unified' && scene.unifiedPalette) {
```
I feel like two checks is a little more than necessary -- this might be easier if we kept a `previousScene` but that didn't really feel worth it...

cycomachead commented on src/objects.js:
```diff
@@ -2435,29 +2410,18 @@ SpriteMorph.prototype.blockTemplates = function (category) {
 
-    // for debugging: ///////////////
-
+        // for debugging: ///////////////
```
the spacing stuff here is all just consistency.

cycomachead commented on src/objects.js:
```diff
@@ -2695,84 +2631,24 @@ SpriteMorph.prototype.blockTemplates = function (category) {
-        blocks.push(button);
+    } else if (category === 'variables') {
 
+        blocks.push(this.makeVariableButton());
```
Again, not really needed, but it makes it more clear exactly what should be in the palette when reading. :)

jmoenig replied:
I like it!

cycomachead commented on src/objects.js:
```diff
@@ -2837,55 +2710,162 @@ SpriteMorph.prototype.blockTemplates = function (category) {
+    return txt;
+}
+
+SpriteMorph.prototype.helpMenu = function () {
```
I don't know if this really belongs to a `SpriteMorph`, but this seems like a useful utility function throughout snap... there's a few other places where this is the entire context menu, I think.

jmoenig replied:
yeah, this is an excellent idea!

cycomachead reviewed (commented):
Actually I've noticed a couple small things I'll need to look at tomorrow.

* some how the `make a block button` can appear twice when toggling modes. (Not sure why...I think the cache is being cleared?)
* Stage local blocks appear twice sometimes...at maybe a weird caching issue?

cycomachead commented:
OK fixed the last two issues. One was cache pollution (hopefully the fix isn't too clever),  and the second was a subtle double-inclusion of blocks in the variables palette. 

It's worth noting that in the unified palette, the ordering of custom blocks is slightly different than from the plain 'variables' palette. In the Unified view the blocks are strictly grouped by category (variables, then lists then other). In the variables palette they're still ordered by creation. That's the current behavior, but it'd be pretty easy to adapt.

cycomachead commented:
<img width="248" alt="Screen Shot 2021-06-30 at 3 07 54 AM" src="https://user-images.githubusercontent.com/1505907/123943946-7ac15100-d951-11eb-9d25-5f049bfba40c.png">

cycomachead commented on src/objects.js:
```diff
@@ -2837,55 +2710,161 @@ SpriteMorph.prototype.blockTemplates = function (category) {
+
+    function addCustomBlock(definition) {
+        if (definition.category === category ||
+                (Array.isArray(category) && category.includes(definition.category))) {
```
Since I tweaked the way this worked, I decided to use the native `includes` method, rather than morphic's `contains`.

jmoenig replied:
Haha, `includes()` didn't exist when Morphic was first made (and it still doesn't exist in IE, I guess. Fine with me!

cycomachead commented on src/objects.js:
```diff
@@ -3095,14 +3085,31 @@ SpriteMorph.prototype.freshPalette = function (category) {
+    if (category === 'unified') {
+        // In a Unified Palette custom blocks appear following each category,
+        // but there is only 1 make a block button (at the end).
+        blocks = this.categories.reduce((blocks, category) =>
```
Oh yes, `flatMap` would work well here, but it doesn't exist in non-Chromium Edge (44) which while no longer supported, got patched a year ago... Not sure what date we should draw the line at...

jmoenig commented:
Excellent!

ddgarcia commented:
Thank you thank you Michael for the good work here and Jens for allowing
this…

When the technical dust settles it would be great to go live with this
asap, since we’re actively writing curricula, taking screenshots, and
setting up starter microworld projects for our July 19th PD launch, well
before snapcon.

Again, thank you!!

Dan

On Wed, Jun 30, 2021 at 12:25 AM Jens Mönig ***@***.***>
wrote:

> Excellent!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmoenig/Snap/pull/2853#issuecomment-871281855>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARAIMKC6YVCCOXNBHU6XXLTVLWIDANCNFSM47RNWLCQ>
> .
>
-- 
*Dan Garcia, Ph.D.* 777 Soda Hall <https://eecs.berkeley.edu/about/visiting>,
+1.510.517.4041 <+15105174041>, ***@***.***
Teaching Professor
<https://www2.eecs.berkeley.edu/Faculty/Homepages/garcia.html>,    UC
Berkeley <https://www.berkeley.edu/>, EECS dept <https://eecs.berkeley.edu/>,
 cs.berkeley.edu/~ddgarcia
Pronouns: He/Him/His <https://www.mypronouns.org/he-him>,   SIGCSE
Vice-Chair <https://sigcse.org/sigcse/about/board.html>,   BJC Co-PI
<http://bjc.berkeley.edu/>,   Snap*!* evangelist <http://snap.berkeley.edu/>
ACM <http://www.acm.org/> Distinguished Educator
<https://awards.acm.org/distinguished-members/award-winners> & Speaker
<http://speakers.acm.org/speakers/garcia_11513>, CSforCA
<http://access-ca.org/csforca/>, CSforALL <https://www.csforall.org/>,
CSTA-GG <https://sites.google.com/site/ggcsta/>, LinkedIn
<http://www.linkedin.com/in/ddgarcia>

jmoenig commented:
Dan, Michael, can we maybe think about a way to "go live" with this in a sort of public beta of v7? I'm not at all sure all of this will be ready to be officially released in this short a time with so many loose ends (on my side) and so little testing. Perhaps using the dev hosting url could be an option?

bromagosa commented:
@cycomachead one small detail: when clicking on a category, the scrollbar doesn't scroll to where it should. It instead just stays where it was before you clicked.

jmoenig commented:
yeah, there are a bunch of little details we have to resolve, none of them hard, but they all require some time. Here are my own notes about them:

* fix scrolling in unified palette (update the scroll-bar, should be easy)
* add scrolling animation (Jens can take care of that)
* change position of the unified palette setting, perhaps save the preference in the browser (not in the project, or maybe in the project?)
* make sure it works with additional / less categories
* make sure it works across several sprites (does it cache the unified palette per sprite?)
* find a way to switch back from search bar to unified palette
* support hiding / showing primitives / categories of primitives in unified palette mode

ddgarcia commented:
Public beta can work, yes. We can be making projects, doing screenshots,
and saving starter links.  At some point we’d switch the links to point to
the non beta version, but that could happen after the PD…

Dan

On Thu, Jul 1, 2021 at 12:17 AM Jens Mönig ***@***.***> wrote:

> yeah, there are a bunch of little details we have to resolve, none of them
> hard, but they all require some time. Here are my own notes about them:
>
>    - fix scrolling in unified palette (update the scroll-bar, should be
>    easy)
>    - add scrolling animation (Jens can take care of that)
>    - change position of the unified palette setting, perhaps save the
>    preference in the browser (not in the project, or maybe in the project?)
>    - make sure it works with additional / less categories
>    - make sure it works across several sprites (does it cache the unified
>    palette per sprite?)
>    - find a way to switch back from search bar to unified palette
>    - support hiding / showing primitives / categories of primitives in
>    unified palette mode
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmoenig/Snap/pull/2853#issuecomment-872115655>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARAIMOCUGXN2MVVUY7OJODTVQ6E7ANCNFSM47RNWLCQ>
> .
>
-- 
*Dan Garcia, Ph.D.* 777 Soda Hall <https://eecs.berkeley.edu/about/visiting>,
+1.510.517.4041 <+15105174041>, ***@***.***
Teaching Professor
<https://www2.eecs.berkeley.edu/Faculty/Homepages/garcia.html>,    UC
Berkeley <https://www.berkeley.edu/>, EECS dept <https://eecs.berkeley.edu/>,
 cs.berkeley.edu/~ddgarcia
Pronouns: He/Him/His <https://www.mypronouns.org/he-him>,   SIGCSE
Vice-Chair <https://sigcse.org/sigcse/about/board.html>,   BJC Co-PI
<http://bjc.berkeley.edu/>,   Snap*!* evangelist <http://snap.berkeley.edu/>
ACM <http://www.acm.org/> Distinguished Educator
<https://awards.acm.org/distinguished-members/award-winners> & Speaker
<http://speakers.acm.org/speakers/garcia_11513>, CSforCA
<http://access-ca.org/csforca/>, CSforALL <https://www.csforall.org/>,
CSTA-GG <https://sites.google.com/site/ggcsta/>, LinkedIn
<http://www.linkedin.com/in/ddgarcia>

cycomachead commented:
> @cycomachead one small detail: when clicking on a category, the scrollbar doesn't scroll to where it should. It instead just stays where it was before you clicked.

Thanks! Fixed. :)

> * fix scrolling in unified palette (update the scroll-bar, should be easy)
Fixed this.
> * add scrolling animation (Jens can take care of that)
Thanks! I tried reading through Frame/ScrollFrame and didn't see something that would handle this. If there's an example I can look at, I'd try to fix this myself...
> * change position of the unified palette setting, perhaps save the preference in the browser (not in the project, or maybe in the project?)
Where should it be? The setting *needs* to be with the project for the BJC curriculum. Right now it's stored on the scene like we talked about a few weeks ago...so each scene can have its own palette style.
> * make sure it works with additional / less categories
Is this really happening soon? I think we should be all set though, since I've been referencing `this.categories`.
> * make sure it works across several sprites (does it cache the unified palette per sprite?)
Based on my testing, yes -- it works fine with multiple sprites. The cache is per-sprite, just like today. (I am really just reusing the `paletteCache`. There's no particular caching changes specific to a unified view.
> * find a way to switch back from search bar to unified palette
This works just fine. :) 
> * support hiding / showing primitives / categories of primitives in unified palette mode
Hiding/showing primitives works exactly the same as before, and the list of primitives hidden is pre-project just like it is today.  Hiding a list of primitives by category would be useful, but I think it is a separate feature — and not really specific to the unified palette. 

Keep in mind the point of the unified palette would be primarily to select a fairly limited set of blocks. The intention is *not* to have folks scroll through hundreds of blocks -- though certainly there's some use for that and it works just fine.
The one limitation when working with hundreds of blocks, is that it would be great if the IDE (or scene?) could remember you scroll position in the palette. That way when switching sprites or toggling searching there's less jumping around. To be clear, those things occur today! Switching sprites and toggling searching always scroll you back up to the top of the palette. So I definitely think this is a follow up pull request. 😄

jmoenig commented:
this all sounds great to me!

jmoenig commented on src/gui.js:
```diff
@@ -1255,6 +1256,27 @@ IDE_Morph.prototype.createCategories = function () {
+    if (this.scene.unifiedPalette) {
+        categorySelectionAction = scrollToCategory;
+    } else {
+        categorySelectionAction = changePallete;
```
umm, not sure (I always have to look this up) but isn't it spelled "palette"?

jmoenig replied:
also, how did I start a "review", I just wanted to comment... haha

cycomachead replied:
Ooops! One L. :) at least autocomplete works! 

And haha, GitHub’s default thing is to leave a review now.

jmoenig commented:
I've added scrolling animations.
There are quite a few issues and bugs happening when I stress-test this feature, I think I'll need to work on it some more (not the animation, the unified palette, especially the workflow and interactions with search and switching to other sprites is tricky). But it's also kinda neat, I like it!

cycomachead commented on src/objects.js:
```diff
@@ -2837,55 +2710,161 @@ SpriteMorph.prototype.blockTemplates = function (category) {
+    return menu;
+}
+
+// returns an array alock templates for a selected category.
```
ooops...wild typo there. 'an array of block templates ...'

jmoenig commented:
Thanks! I've just changed it directly, since I've already merged scenes into main by now (there's no turning back now, haha!)
